theano/compile/compilelock.py

```diff
@@ -0,0 +1,67 @@
+"""
+Locking mechanism to ensure no two compilations occur simultaneously
+in the same compilation directory (which can cause crashes).
+"""
+import os
+import typing
+from contextlib import contextmanager
+
+import filelock
+
+from theano import config
+
+
+__all__ = [
+    "force_unlock",
+    "lock_ctx",
+]
+
+
+_locks = {}
+
+
+def force_unlock(lock_dir: os.PathLike):
+    """
+    Forces the release of the lock on a specific directory.
+
+    Parameters
+    ----------
+    lock_dir : os.PathLike
+        Path to a directory that was locked with `lock_ctx`.
+    """
+    fl, sl = _locks[lock_dir]
+    fl.release(force=True)
+    sl.release(force=True)
+
+
+@contextmanager
+def lock_ctx(lock_dir=None, *, timeout: typing.Optional[float] = -1):
+    """Context manager that wraps around FileLock and SoftFileLock from filelock package.
+
+    Parameters
+    ----------
+    lock_dir : str
+        A directory for which to acquire the lock.
+        Defaults to the config.compiledir.
+    timeout : float
+        Timeout in seconds for waiting in lock acquisition.
+        Defaults to config.compile__timeout.
+    """
+    if lock_dir is None:
+        lock_dir = config.compiledir
+    if timeout == -1:
+        timeout = config.compile__timeout
+    elif timeout is not None or timeout <= 0:
+        raise ValueError(f"Timeout parameter must be None or positive. Got {timeout}.")
+
+    # locks are kept in a dictionary to account for changing compiledirs
+    if lock_dir not in _locks:
+        fl = filelock.FileLock(os.path.join(lock_dir, ".lock"))
+        sl = filelock.SoftFileLock(os.path.join(lock_dir, ".softlock"))
```

Comment:
I don't think we need to use _both_ `FileLock` and `SoftFileLock`; it looks like [`filelock` chooses the appropriate one itself](https://github.com/benediktschmitt/py-filelock/blob/b30bdc4fb998f5f4350257235eb51147f9e81862/filelock.py#L448) on module load.

Reply:
Right, well spotted!